common/djangoapps/student/models.py

```diff
@@ -1442,6 +1444,8 @@ def enrollment_mode_for_user(cls, user, course_id):
             and is_active is whether the enrollment is active.
         Returns (None, None) if the courseenrollment record does not exist.
         """
+        if user.is_anonymous():
```

Comment:
1. I'm not a fan of allowing functions to be loose when handling parameters. I'll try fixing the one above to require a user to be passed (there were strange test failures that I gave up debugging but I'll try again).

2. Interesting idea. The logic is a little more complicated in ``_get_enrollment_state`` since it returns two values, but it simplifies two callers.